### ISSUE TITLE
Music issues [Suggestion] 

### ISSUE THREAD

**Issue by ChristopherJ123** [quiqueck/BetterEnd#104]
### Description

I noticed that betterend music would change depending on the biome you're in, and each biome has their own unique ambient sounds. I suggest that betterend music would keep playing in the background even if we go to another biome until the music ends but the ambient sounds would change on the biome. This is done to keep immersion while exploring and also resemble Vanilla music system. A noticeable case of this issue is when you're mining in the jade caves biome and you stepped into another biome chunk the music would instantly stop and switch to other music, then it happens again and again.

**Comment by OpenBagTwo** [comment]
The code that covers this behavior can be found in the `MusicTrackerMixin`:

https://github.com/quiqueck/BetterEnd/blob/a5bfe04ed1c239f2c0765face7e16319b8414a3c/src/main/java/org/betterx/betterend/mixin/client/MusicTrackerMixin.java#L42-L83

Trying to follow the logic of that giant `if` statement, especially when combined with this method:

https://github.com/quiqueck/BetterEnd/blob/a5bfe04ed1c239f2c0765face7e16319b8414a3c/src/main/java/org/betterx/betterend/mixin/client/MusicTrackerMixin.java#L93-L98

is extremely difficult, but going through a few different versions of the code, all the way back to [the first version of this mixin](https://github.com/paulevsGitch/BetterEnd/blob/6362c2f85ada34778910da1a8f90b90ebf51c185/src/main/java/ru/betterend/mixin/client/MusicTrackerMixin.java)

the intended behavior appears to be:

1. **IF** the config specifies to play biome music, and
2. **IF** the situation calls for music, and
3. **IF** the there is already music playing. and
4. **IF** the player is in the end, and
5. **IF** the situational music is not already playing
6. **IF** something about the music or sound event indicates that it's a track that should interrupt other tracks.

Then, perform a slow fade out of the current track, over the course of 20 seconds.

Once volume hits zero, stop music.

Once there's no longer any music playing, count down a random number of ticks (2.5 minutes average **I think**) before playing the end biome music.

---

Weirdly, I'm not actually sure how the block on L75 (the part that actually plays the music) could get triggered, as if `this.currentmusic == null` then the second check fails, and we skip right to L79. Yes, we have `this.currentMusic = null;` on L73, but `this.nextSongDelay--` is only going to get one opportunity to get called, and then never called again (unless, of course, `this.nextSongDelay` is _initially set_ to zero, which would mean that the "choose a random time delay between songs" code isn't working, which, you know, might be.

So to @paulevsGitch I would ask--did this code _ever_ work, or am I reading something wrong? I saw that for a while the class had a `println` debug statement trackling the volume level fade for a while, so I'm guessing it gave you a little trouble. At the same time, I'll keep it :100: that I've rewritten this writeup three times as I keep on figuring out something new parts of what this code is doing.

**Comment by paulevsGitch** [comment]
> did this code ever work

Yes, when that code was made in 1.16.5 it worked fine. Original method looked like [this](https://github.com/paulevsGitch/BetterEnd/blob/1.16.5/src/main/java/ru/betterend/mixin/client/MusicTrackerMixin.java#L41-L82). Its logic worked like this:
1. IF music blending is on
2. IF music is not null
3. IF music volume is larger that zero
4. IF we are in End
5. IF music require changing:
    - Music location is different
    - Music can be changed

If all conditions are fine it calculates new music volume:
1. IF volume is larger than zero decrease it on calculated delta and set
2. IF it is equal to zero set it to one and stop music

If music is null it will start playing a new song.
Original code was simple, and worked fine. However during time it was changed, together with MC code, and since, probably, 1.18 it stop working as expected

**Comment by OpenBagTwo** [comment]
Thanks! That makes a lot of sense, especially now that I'm looking at the vanilla `MusicManager` implementation (which at least explains how music can _start_ playing).

I'm doing some empirical testing now--I was curious what would happen if `blendBiomeMusic` were set to `false` in the config. It looks like the answer is: music goes to zero the second you exit an end biome, and, say, enter the void to fly to a different island. That is _not_ what I would have expected, unless the void has its own interrupting trigger.

Before proposing any changes I do want to make sure that everyone is firm on the _expected_ behavior, according to each of the two options. My two cents would be:

- **If music blending is off**, then let any situational music play to completion before a new track starts
- **If music blending is on**, then use the fade out behavior you just described.

So that would mean people like @ChristopherJ123 (and myself) would likely want to set the config value to `false`.

**Comment by OpenBagTwo** [comment]
Currently in Minecraft 1.21.6, C418 - The End is **hard-coded** to play (and to stop playing any existing music) any time the player is in The End dimension (and the dragon isn't running around).

This is now fixed with a Mixin in LighterEnd https://github.com/OpenBagTwo/LighterEnd/commit/042c378f9620f9f6126c16d969604b26e8f27157

Just in case anyone is curious.
